Plugins/Json/src/MaterialJsonConverter.cpp

```diff
@@ -674,3 +676,139 @@ void Acts::from_json(const nlohmann::json& j, volumeMaterialPointer& material) {
+      } else if (bUtility.binningData()[0u].binvalue == binZ) {
+        // Turn to Phi-Z - swap needed
+        BinUtility nbUtility(1u, -M_PI, M_PI, closed, binPhi);
+        nbUtility += bUtility;
+        bUtility = std::move(nbUtility);
+        swapped = true;
+      } else {
+        std::runtime_error("Unsupported binning for Detray");
+      }
+    } else if (bUtility.dimensions() == 2u &&
+               bUtility.binningData()[0u].binvalue == binZ &&
+               bUtility.binningData()[1u].binvalue == binPhi) {
+      BinUtility nbUtility(bUtility.binningData()[1u]);
+      nbUtility += bUtility.binningData()[0u];
+      bUtility = std::move(nbUtility);
+      swapped = true;
+    }
+
+    BinningValue bVal0 = bUtility.binningData()[0u].binvalue;
+    BinningValue bVal1 = bUtility.binningData()[1u].binvalue;
+
+    // Translate into grid index type
+    int gridIndexType = 0;
+    if (bVal0 == binR && bVal1 == binPhi) {
+      gridIndexType = 0;
+    } else if (bVal0 == binPhi && bVal1 == binZ) {
+      gridIndexType = 3;
+    } else if (bVal0 == binX && bVal1 == binY) {
+      gridIndexType = 2;
+    } else {
+      std::runtime_error("Unsupported binning for Detray");
+    }
+    // Convert the axes
+    nlohmann::json jAxes = toJsonDetray(bUtility, surface);
+    // Create  a grid index, i.e. type, index tuple
+    nlohmann::json jGridLink;
+    jGridLink["type"] = gridIndexType;
+    std::size_t gridIndex = 0;
+    if (gridLink.find(gridIndexType) != gridLink.end()) {
+      std::size_t& fGridIndex = gridLink[gridIndex];
+      gridIndex = fGridIndex;
+      fGridIndex++;
+    } else {
+      gridLink[gridIndexType] = 1;
+    }
+    jGridLink["index"] = gridIndex;
+
+    // The grid data
+    jSurfaceMaterial["axes"] = jAxes;
+    jSurfaceMaterial["grid_link"] = jGridLink;
+    jSurfaceMaterial["owner_link"] = surfaceIndex;
+
+    // The bins to be filled
+    nlohmann::json jBins;
+    auto materialMatrix = binnedMaterial->fullMaterial();
+    for (std::size_t ib1 = 0; ib1 < materialMatrix.size(); ++ib1) {
+      for (std::size_t ib0 = 0; ib0 < materialMatrix[0u].size(); ++ib0) {
+        nlohmann::json jBin;
+        // Look up the material slab
+        MaterialSlab slab = materialMatrix[ib1][ib0];
+        // Translate into a local bin
+        std::size_t lb0 = swapped ? ib1 : ib0;
+        std::size_t lb1 = swapped ? ib0 : ib1;
+        jBin["loc_index"] = std::array<std::size_t, 2u>{lb0, lb1};
+
+        const Material& material = slab.material();
+        // The content
+        nlohmann::json jContent;
+        jContent["thickness"] = slab.thickness();
+        // The actual material
+        nlohmann::json jMaterialParams;
+        if (slab.thickness() > 0.) {
+          jMaterialParams["params"] = std::vector<ActsScalar>{
+              material.X0(),          material.L0(),          material.Ar(),
+              material.Z(),           material.massDensity(), 0,
```

Comment:
I think, the material state and molar density are swapped. This is what the detray homogeneous material writer does:
```

        mat_data.params = {mat.X0(),
                           mat.L0(),
                           mat.Ar(),
                           mat.Z(),
                           mat.mass_density(),
                           mat.molar_density(),
                           static_cast<real_io>(mat.state())};
```

Reply:
I see, switching it now. What is the `state` here?

Reply:
I believe its an enum that defines solid, liquid etc, but I don't really know what its used for